### ISSUE TITLE
IO.bracketCase possibly broken?

### ISSUE THREAD

**Issue by smackaj** [arrow-kt/arrow#1359]
I hope I don't just understand `IO.bracketCase(release,use)` wrong. My understanding is that when running the resulting IO action, `release` should always be run after running `use`, regardless of the result of `use`. Moreover, when `use: (A) -> IO<B>` results in `IO.raiseError`, I would expect that `release:(A,ExitCase)->IO<Unit>` enters the `ExitCase.Error` branch.

**Issue:** The following test fails:
`
    @Test
    internal fun `bracketCase works as expected`() {

        val transaction: Transaction = mock{
            on{ commit(Unit)} doReturn IO.unit
            on{ rollback(Unit)} doReturn IO.unit
        }

        val e: Exception = IllegalStateException("Hi!")

        val use: (Transaction) -> IO<Int> = { IO.raiseError(e) }
        val release: (Transaction,ExitCase<Throwable>)-> IO<Unit> =
            {t, exitCase -> when(exitCase){
                is ExitCase.Completed -> t.commit(Unit)
                else -> t.rollback(Unit)
            }}

        val result: Either<Throwable,Int> = IO.just(transaction).bracketCase(release,use).attempt().unsafeRunSync()

        assertThat(result).isEqualTo(e.left())

        verify(transaction, never()).commit(Unit)
        verify(transaction, times(1)).rollback(Unit)
    }`

Error message is: _Zero_ interactions with this mock (transaction).
In other words, `release` is never executed. 
In other similar tests I am consistently unable to ever enter the `ExitCase.Error` or `ExitCase.Cancelled` branches.

I posted a [related question on stackoverflow](https://stackoverflow.com/questions/55128770/kotlin-arrow-io-transactions-with-io-bracketcase) yesterday and I hope posting here right afterwards is not annoying. Still, I am more and more convinced that the above behaviour just cannot be right so there you are. :)

**Comment by pakoito** [comment]
There's a chance Attempt is the broken one. Try without it.

**Comment by smackaj** [comment]
No, that didn't change the behaviour.

**Comment by nomisRev** [comment]
Hi @smackaj,

With what version of Arrow did you see this behavior? I could not reproduce this, the following tests passes for me.

```kotlin
import arrow.effects.*
import arrow.effects.extensions.io.concurrent.concurrent
import arrow.effects.extensions.io.monad.flatMap
import arrow.effects.handleErrorWith
import arrow.effects.typeclasses.ExitCase
import io.kotlintest.shouldBe

val e: Exception = IllegalStateException("Hi!")

fun main() {
  Promise<ForIO, ExitCase<Throwable>>(IO.concurrent()).flatMap { latch ->
    IO.unit //Fake resource
      .bracketCase(use = {
        IO.raiseError<String>(e) //Immediately blow up while consuming resource
      }, release = { _, exitCase ->
        latch.complete(exitCase)      //Store the ExitCase result in the promise
      })
      .handleErrorWith { latch.get() } //Replace result of `use` with result from latch.
  }
    .unsafeRunSync() shouldBe ExitCase.Error(e) //Assert correct result
}
```

There were some issues with `Bracket` in the past but I thought they were already resolved so could you double check version? Your understanding of `Bracket` seems correct.

Thanks for reporting. And my apologies for the late response.

**Comment by nomisRev** [comment]
@smackaj did the proposed solution for work for you?
